metals/src/main/scala/scala/meta/internal/metals/FileWatcher.scala

```diff
@@ -49,25 +49,26 @@ final class FileWatcher(
   }
 
   def restart(): Unit = {
-    val directoriesToWatch = new util.ArrayList[Path]()
+    val sourcesItemsToWatch = new util.ArrayList[Path]()
     val createdSourceDirectories = new util.ArrayList[AbsolutePath]()
-    def watch(dir: AbsolutePath, isSourceDirectory: Boolean): Unit = {
-      if (!dir.isDirectory) {
-        dir.createDirectories()
-        if (isSourceDirectory) createdSourceDirectories.add(dir)
+    def watch(path: AbsolutePath, isSourceItem: Boolean): Unit = {
+      if (!path.isDirectory && !path.isFile) {
+        path.createDirectories()
+        if (isSourceItem) createdSourceDirectories.add(path)
       }
-      directoriesToWatch.add(dir.toNIO)
+      if (path.isFile) sourcesItemsToWatch.add(path.toNIO.getParent())
```

Comment:
let's hope this won't become a bottleneck

Reply:
Added a separate file watcher that watches parent dir and filters on the known files.